lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,434 @@
+  const Scalar minInd = data.getMin()[1];
+  const Scalar maxInd = data.getMax()[1];
+
+  /* Add text description */
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = (k + 1.0) + 0.08;
+    data(k, 1) = values[k];
+  }
+
+  Description names(values.getDimension());
+  for(UnsignedInteger i = 0; i < values.getDimension(); ++i)
+  {
+    OSS oss;
+    oss << "X" << i + 1;
+    names[i] = String(oss);
+  }
+
+  Text text(data, names, "right");
+  text.setColor("black");
+  graph.add(text);
+
+  // Set bounding box
+  Point lowerBound(2);
+  lowerBound[0] = 0.0;
+  if (minInd < 0)
+    lowerBound[1] = 1.1 * minInd;
+  else
+    lowerBound[1] = 0.9 * minInd;
+
+  Point upperBound(2);
+  upperBound[0] = values.getDimension() + 1.0 ;
+
+  if (maxInd > 0)
+    upperBound[1] = 1.1 * maxInd;
+  else
+    upperBound[1] = 0.9 * maxInd;
+
+  graph.setBoundingBox(Interval(lowerBound, upperBound));
+
+  return graph;
+}
+
+
+/** Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawHSICIndices()
+{
+  String title = "HSIC indices";
+  Graph graph = drawValues(getHSICIndices(), title);
+  return graph;
+}
+
+/** Draw the R2-HSIC indices */
+Graph HSICEstimatorImplementation::drawR2HSICIndices()
+{
+  String title = "R2-HSIC indices";
+  Graph graph = drawValues(getR2HSICIndices(), title);
+  return graph;
+}
+
+/** Draw the p-values */
+Graph HSICEstimatorImplementation::drawPValuesPermutation()
+{
+  String title = "p-values by permutation";
+  Graph graph = drawValues(getPValuesPermutation(), title);
+  return graph;
+}
+
+/** Get the covariance list */
+CovarianceModelCollection HSICEstimatorImplementation::getCovarianceList() const
+{
+  return covarianceList_;
+}
+
+/** Set the covariance list: dimension is input dimension + 1 */
+void HSICEstimatorImplementation::setCovarianceList(const CovarianceModelCollection & coll)
+{
+  covarianceList_ = coll ;
+  resetIndices();
+}
+
+/** Get the input sample */
+Sample HSICEstimatorImplementation::getInputSample() const
+{
+  return inputSample_;
+}
+
+/** Set the input sample */
+void HSICEstimatorImplementation::setInputSample(const Sample & inputSample)
+{
+  inputSample_ = inputSample;
+  resetIndices();
+}
+
+/** Get the output sample */
+Sample HSICEstimatorImplementation::getOutputSample() const
+{
+  return outputSample_;
+}
+
+/**Set the output sample: must be of dimension one */
+void HSICEstimatorImplementation::setOutputSample(const Sample & outputSample)
+{
+  if(outputSample.getDimension() != 1)
+  {
+    throw NotYetImplementedException(HERE) << "Dimension of output sample should be 1.";
+  }
+
+  outputSample_ = outputSample;
+  resetIndices();
+}
+
+/** Get the weight function */
+Function HSICEstimatorImplementation::getWeightFunction() const
+{
+  return weightFunction_;
+}
+
+/** Set the weight function: user should check properties! */
+void HSICEstimatorImplementation::setWeightFunction(const Function & weightFunction)
```

Comment:
f I understand correctly, now `HSICEstimatorImplementation` should behave as a `HSICEstimatorGlobalSensitivity` estimator?

Reply:
Then, should I remove the attribute as well? That would make sense.

Reply:
> If I understand correctly, now `HSICEstimatorImplementation` should behave as a `HSICEstimatorGlobalSensitivity` estimator?
> Then, should I remove the attribute as well? That would make sense.

No, `HSICEstimatorImplementation` will still essentially behave like `HSICEstimatorConditionalSensitivity`. It will just be deprived of the method `setWeightFunction`. The weight function can still be specified in the constructor of `HSICEstimator`, so this is a really minor difference.

Basically, I want `HSICEstimatorImplementation` to behave like `HSICEstimatorConditionalSensitivity`, but I do not want `HSICEstimatorGlobalSensitivity` and `HSICEstimatorTargetSensitivity` to inherit `setWeightFunction`. The only solution I can see is the one I suggested: move  `setWeightFunction` from `HSICEstimatorImplementation` to `HSICEstimatorConditionalSensitivity`.

When you construct an object through the interface class (which is possible even though it does not make much sense for a Python user), it must yield a valid object. For example, when you call `dist = ot.Distribution()`, `dist` actually behaves like a `Uniform()` - except specific accessors like `setA()` and `setB()` are unavailable.

To me, `HSICEstimator` is analoguous to `Distribution` and `HSICEstimatorConditionalSensitivity` is analogous to `Uniform`: ``dist = ot.Distribution()`` is roughly equivalent to ``dist = ot.Uniform()``.
However, the accessor `setWeightFunction` is analoguous to `setA` and `setB`: the former is only available in `HSICEstimatorConditionalSensitivity`, the latter are only avaialbe in `Uniform`.